controllers/reclaimspacejob_controller.go

```diff
@@ -18,45 +18,332 @@ package controllers
+		return ctrl.Result{}, nil
+	}
+
+	if rsJob.Status.Result != "" {
+		logger.Info(fmt.Sprintf("ReclaimSpaceJob is already in %q state, exiting reconcile",
+			rsJob.Status.Result))
+		// since result is already set, just dequeue
+		return ctrl.Result{}, nil
+	}
+
+	err = validateReclaimSpaceJobSpec(rsJob)
+	if err != nil {
+		logger.Error(err, "failed to validate ReclaimSpaceJob.Spec")
+
+		rsJob.Status.Result = csiaddonsv1alpha1.OperationResultFailed
+		rsJob.Status.Message = fmt.Sprintf("failed to validate ReclaimSpaceJob.Spec: %v", err)
+		if statusErr := r.Client.Status().Update(ctx, rsJob); statusErr != nil {
+			logger.Error(err, "failed to update status")
+			return ctrl.Result{}, statusErr
+		}
+
+		// invalid parameter, do not requeue.
+		return ctrl.Result{}, nil
+	}
+
+	err = r.reconcile(
+		ctx,
+		&logger,
+		&rsJob.Spec,
+		&rsJob.Status,
+		req.Namespace,
+	)
+
+	if rsJob.Status.Result == "" && rsJob.Status.Retries == rsJob.Spec.BackoffLimit {
+		logger.Info("Maximum retry limit reached")
+		rsJob.Status.Result = csiaddonsv1alpha1.OperationResultFailed
+		rsJob.Status.Message = "Maximum retry limit reached"
+		rsJob.Status.CompletionTime = &v1.Time{Time: time.Now()}
+	}
+
+	if statusErr := r.Client.Status().Update(ctx, rsJob); statusErr != nil {
+		logger.Error(statusErr, "failed to update status")
 
-	// TODO(user): your logic here
+		return ctrl.Result{}, statusErr
+	}
 
-	return ctrl.Result{}, nil
+	if rsJob.Status.Result != "" {
+		// since result is already set, just dequeue
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, err
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ReclaimSpaceJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&csiaddonsv1alpha1.ReclaimSpaceJob{}).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
 		Complete(r)
 }
+
+func (r *ReclaimSpaceJobReconciler) reconcile(
+	ctx context.Context,
+	logger *logr.Logger,
+	spec *csiaddonsv1alpha1.ReclaimSpaceJobSpec,
+	status *csiaddonsv1alpha1.ReclaimSpaceJobStatus,
+	namespace string) error {
+
+	if status.StartTime == nil {
+		status.StartTime = &v1.Time{Time: time.Now()}
+	} else {
+		// not first reconcile, increment retries
+		status.Retries++
+	}
+
+	if time.Now().After(status.StartTime.Time.Add(time.Second * time.Duration(spec.RetryDeadlineSeconds))) {
+		status.Result = csiaddonsv1alpha1.OperationResultFailed
+		status.Message = "Time limit reached"
+		status.CompletionTime = &v1.Time{Time: time.Now()}
+
+		return nil
+	}
+
+	driverName, pvName, nodeID, err := r.getTargetDetails(ctx, logger, spec.Target, namespace)
+	if err != nil {
+		logger.Error(err, "Failed to get target details")
+		status.Message = fmt.Sprintf("Failed to get target details: %v", err)
+		//setFailedStatus
+
+		return err
+	}
+
+	controllerFound, controllerReclaimedSpace, err := r.controllerReclaimSpace(ctx, logger, driverName, pvName)
```

Comment:
the better order would be node first, then controller. In case the node zero's out reclaimed space, the controller can do more reclaiming (in case of Ceph, at least)